ddtrace/internal/_encoding.pyx

```diff
@@ -46,120 +47,327 @@ cdef inline int PyBytesLike_Check(object o):
-    """
+    def __init__(self):
+        self._index = {"": 0}
+        self.insert("")
+
+    cpdef insert(self, object string):
+        pass
+
+    cpdef index(self, object string):
+        cdef int index
+
+        if not string:
+            return 0
+        
+        try:
+            return self._index[string]
+        except KeyError:
+            index = len(self)
+            self._index[string] = index
+            self.insert(string)
+            return index
+
+    def __len__(self):
+        return len(self._index)
+
+    def __contains__(self, object string):
+        return string in self._index
+
+
+cdef class MsgpackStringTable(StringTable):
     cdef msgpack_packer pk
-    cdef object _default
-    cdef const char *encoding
-    cdef const char *unicode_errors
+    cdef int max_size
+    cdef int _sp
+    cdef object _lock
 
-    def __cinit__(self):
-        cdef int buf_size = 1024*1024
-        self.pk.buf = <char*> PyMem_Malloc(buf_size)
+    def __init__(self, max_size):
+        self.pk.buf = <char*> PyMem_Malloc(max_size)
         if self.pk.buf == NULL:
             raise MemoryError("Unable to allocate internal buffer.")
-        self.pk.buf_size = buf_size
-        self.pk.length = 0
+        self.max_size = self.pk.buf_size = max_size
+        self.pk.length = 6
+        self._sp = 0
+        self._lock = threading.Lock()
+        super(MsgpackStringTable, self).__init__()
+
+    def __dealloc__(self):
+        PyMem_Free(self.pk.buf)
+        self.pk.buf = NULL
 
-    def __init__(self, default=None):
-        if default is not None:
-            if not PyCallable_Check(default):
-                raise TypeError("default must be a callable.")
-        self._default = default
+    cpdef insert(self, object string):
+        cdef int ret
+
+        ret = pack_text(&self.pk, string)
+        if ret != 0:
+            raise RuntimeError("Failed to add string to msgpack string table")
+
+    cpdef savepoint(self):
+        self._sp = self.pk.length
+
+    cpdef rollback(self):
+        if self._sp > 0:
+            self.pk.length = self._sp
+
+    cpdef get_bytes(self):
+        cdef int l = len(self._index)
+        cdef int offset = 6 - array_prefix_size(l)
+        cdef int old_pos = self.pk.length
+        
+        with self._lock:
+            # Update table size prefix
+            self.pk.length = offset
+            msgpack_pack_array(&self.pk, l)
+            # Add root array size prefix
+            self.pk.length = offset = offset - 1
+            msgpack_pack_array(&self.pk, 2)
+            self.pk.length = old_pos
+        
+        return PyBytes_FromStringAndSize(self.pk.buf + offset, self.pk.length - offset)
+
+    @property
+    def size(self):
+        return self.pk.length - 5 + array_prefix_size(len(self._index))
+
+    cpdef append_raw(self, object src, Py_ssize_t size):
+        cdef int res
+        with self._lock:
+            assert self.size + size <= self.max_size
```

Comment:
Should we have an assertion here? Maybe a conditional instead that raises an error?

Reply:
Since `append_raw` is something that we call internally, there should already be guarantees that we are not overflowing the buffer (the checks are triggered by the `put` method). So this is just a sanity check more than anything, hence why I've used an assert. But we could make this into an exception of course 🙂